doc/gnpsi/gnpsi_hld.md

```diff
@@ -0,0 +1,464 @@
+*   `/server`: helper to start and stop relay server and the stats thread, also write to appl_state_db
+*   `/utils`: stats util, credential util and authz logger util.
+
+
+#### Alternative
+
+![alternative design without local loopback](images/image6.png "alternative")
+
+
+**Advantages:**
+
+
+
+*   Avoids the overhead of writing and reading from the local loopback IP.
+
+**Disadvantages:**
+
+
+
+*   Introduces architectural complexities.
+*   Difficult to upstream and maintain the sFlow open-source C code.
+
+Given that the overhead of using the local loopback collector is negligible, we opted for local loopback collector solution instead of integrated processes.
+
+
+### 7. Configuration and management
+
+
+gNMI path
+
+These paths controls the enablement of the server and gRPC port for gNPSI relay server 
+
+
+
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/config/enable`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/config/port`
+
+These paths cover the stats collection for gNPSI
+
+
+
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/bytes-sent`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/packets-sent`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/sample-send-error`
+
+
+Server flag
+
+
+*   `gnpsi_grpc_port`: gRPC port for the gNPSI Server
+*   `gnpsi_max_clients`: the max number of clients that can connect to gNPSI server at same time
+*   `udp_port` UDP port to read sFlow samples. Switch would configure localhost and this port as sFlow collector.
+
+
+#### 7.1 Config DB Enhancements
+
+gNPSI uses `CONFIG_DB`, `APPL_STATE_DB` and `COUNTERS_DB` 
+
+DB schema is shown as below:
+
+
+```
+CONFIG_DB:
+	"GNPSI|global": {
+"admin_state": "ENABLE"|"DISABLE",
+"port": "<port_num>"
+}
+	APPL_STATE_DB:
+		"GNPSI:global": {
+"admin_state": "ENABLE"|"DISABLE",
+"port": "<port_num>"
+}
+	COUNTERS_DB:
+		"COUNTERS:GNPSI:<remote_ip>/<port_num>": {
+			"bytes_sent": "0",
+			"packets_sent": "0",
+			"packets_error": "0"
+}
+```
+
+
+
+
+*   For CONFIG_DB/APP_STATE_DB, admin_state and port are added for configuration purposes.
+*   For COUNTERS_DB, gNPSI added stats path `bytes_sent`, `packets_sent` and `packetserror` for each collector._
+
+
+### 8. Warmboot and Fastboot Design Impact
+
+Since gNPSI does not interact with Switch hardware, it has no impact with respect to warm reboot and fast reboot.
+
+
+### 9. Memory Consumption
+
+Memory consumption is not significant based on collected stats. 
+
+![gNPSI memory usage](images/image7.png "memory usage")
+
+
+(All data is collected on Google-internal device environment)
+
+There is a ~10MB(5%) memory increase after integration of feature (even when configuration is disabled)
+
+
+### 10. Testing Design
+
+
+#### 10.1. Unit Test cases
+
```

Comment:
Can you add a sample gNPSI client commands?

Reply:
There is no specific cli tool at this moment, I assume you can use [grpc_cli](https://github.com/grpc/grpc/blob/master/test/cpp/util/grpc_cli.cc) for this.